### ISSUE TITLE
Terminology

### ISSUE THREAD

**Issue by MartinThoma** [TensorVision/TensorVision#1]
We should try to be consistent in the terminology we use.
## Phases of model development
- Adapt the model parameters to the dataset
  - **train**: Used by Lasagne ([source](http://lasagne.readthedocs.org/en/latest/user/tutorial.html#training-the-model))
  - **fit**: Used with sklearn and R ([source](http://www.r-bloggers.com/evaluating-logistic-regression-models/))
  - **learn**: Makes sense because of learning rate
- Apply the model to data and get the results
  - **predict**: Used by sklearn and Lasagne ([source](http://lasagne.readthedocs.org/en/latest/user/tutorial.html#loss-and-update-expressions))
  - **get_output**: Used by Lasagne ([source](http://lasagne.readthedocs.org/en/latest/user/tutorial.html#loss-and-update-expressions))
  - **evaluate**: Unfortunate, because it could also mean to test how well a certain model is.

I'm not sure what **val** (https://github.com/TensorVision/TensorVision/commit/f3bcd7150003e12684f52e8a5e42c3f7b742ec7c) is... I guess predict?
## Datasets
- A dataset being used to adjust model parameters (**training set**)
- A dataset being used to adjust parameters of the training algorithm while training (**validation set**)
- A dataset being used to adjust parameters of the training algorithm before training (**development set**)
- A dataset being used to evaluate how well the model does with unseen data (**testing set**)
## Misc
- network: A concrete example. Changing a single parameter gives a different network
- architecture: The way data is processed
- model: More general than architecture. That does not necessarily have to be a neural network.
- optimizer: Algorithm which tries to find the minimum of a loss function.
- solver: Same as optimizer. Avoid using this one.

**Comment by MarvinTeichmann** [comment]
https://github.com/TensorVision/TensorVision/commit/f3bcd7150003e12684f52e8a5e42c3f7b742ec7c  phase refers to the kind of data loaded. `val` stands for validation data. However according to definitions in Datasets this should be renamed to `test`.

**Comment by MarvinTeichmann** [comment]
I like the terminology **fit** and **predict**. However as **train** seems to be very common in the CNN community **train** and **predict** might be good. I would also provide a script **evaluate**, which runs the model on the whole training data and provides some nice plots.

**Comment by MartinThoma** [comment]
@MarvinTeichmann Please don't see my current definitions as fixed at the current stage of development. They are only what I currently think the words mean. However, at some point we should fix it (and probably add it to the documentation as a glossary).

I'm happy with "train" and "predict".

> I would also provide a script evaluate, which runs the model on the whole training data and provides some nice plots.

I think figuring out how well a model performs is a crucial part of ML development. There are a couple of ways one can evaluate a trained model, given some data:
1. Accuracy
2. Confusion matrices
3. N images per class which gave the strongest activation
4. Per neuron / filter: Which images gave the strongest activation.
5. Some "cheating" / "maniplation" ones, like having an image of class i and changing it automatically as little as possible to get the network predict class j very strong (with i != j, of course)

As there might be a multitude of possibilities to evaluate a model. We should keep that in mind.
